### PR TITLE
Add support for all queries nested fields, and foreign key

### PR DESCRIPTION
**Pull request by rodrigodc07**
Create namespaces that sync nested field, and fix MBQL->native queries

### CODE REVIEW

**Comment by dacort** [comment]
This is fantastic, thank you. Will take a look at this in the next few days.

**Comment by dacort** [comment]
I'm having trouble running this on some nested tables I have. Receiving a stack overflow error:

```
05-20 09:38:35 ERROR driver.athena :: Error retreiving fields for DB [redacted_db].2017_nestedparquet
java.lang.StackOverflowError
	at java.util.regex.Matcher.replaceFirst(Matcher.java:1000)
	at clojure.string$replace_first.invokeStatic(string.clj:174)
	at clojure.string$replace_first.invoke(string.clj:138)
	at metabase.driver.hive_parser$cool_parser.invokeStatic(hive_parser.clj:20)
	at metabase.driver.hive_parser$cool_parser.invoke(hive_parser.clj:5)
	at metabase.driver.hive_parser$cool_parser.invokeStatic(hive_parser.clj:28)
	at metabase.driver.hive_parser$cool_parser.invoke(hive_parser.clj:5)
        ...
```

It does work with others, so must be an issue with this specific table, will need to dig in a little more.

**Comment by rodrigodc07** [comment]
The stack overflow happens depending on the nesting depth. The parser uses recursion to analyse the schema. To workaround this issue just set the -Xss JVM parameter. Here we use 10m.

**Comment by dacort** [comment]
Hm, ok. I wonder if there's a more performant way to parse the structure?

Do you know what the limit is by default so we could document it? e.g. is it the length of the struct field? The number of fields?

File: src/metabase/driver/athena.clj [hunk 1]
```diff
@@ -69,7 +78,7 @@
     :map        :type/*
     :smallint   :type/Integer
     :string     :type/Text
-    :struct     :type/*
+    :struct     :type/Dictionary
     :timestamp  :type/DateTime
     :tinyint    :type/Integer
     :varchar    :type/Text} database-type))
```
Comment (dacort) [inline src/metabase/driver/athena.clj]: Can/should we pull in the `column->base-type` definitions from `schema_parser`?

I'm not very familiar with Clojure so unsure if this is a valid suggestion.

File: src/metabase/driver/athena.clj [hunk 2]
```diff
@@ -127,23 +138,65 @@
                             database-type))
           :type/*)))
 
+(defn- run-query
+  "Workaround for avoiding the usage of 'advance' jdbc feature that are not implemented by the driver yet.
+   Such as prepare statement"
+  [database query]
+  (log/infof "Running Athena query : '%s'..." query)
+  (try
+    (jdbc/query (sql-jdbc.conn/db->pooled-connection-spec database) (string/replace query ";" " ") {:raw? true})
```
Comment (dacort) [inline src/metabase/driver/athena.clj]: The `replace query ";" " "` function is useful for native SQL queries as well. Is there a way to use this function for those?

File: src/metabase/driver/athena.clj (hunk 1 above)
Comment (rodrigodc07) [reply src/metabase/driver/athena.clj]: Yes, i did it on this commit [https://github.com/dacort/metabase-athena-driver/pull/10/commits/e1fcba3cac8eb3cc93f799f68332bb6ca09d812e](url)

**Comment by nicolasterral** [comment]
What is the strategy to maintain the driver in the future ?

In less than 2 weeks, the code provided by Rodrigo is already broken because of a call to the qualify-and-escape function in the query_processor.

There are some changes in the honeysql-extensions :
https://github.com/metabase/metabase/commit/bafca293295b487d76a1aad1ecea413f22d49b2d#diff-b5d79995c30d3d17cd820cfe745fc99a

Anyway, nice work guys. I can't wait to try this out.

**Comment by dacort** [comment]
@nicolasterral As noted in #12 we certainly need to add tests to this driver.

In addition to basic unit tests, it appears we should also test against different versions of Metabase to ensure compatibility. The driver codebase is certainly in the midst of changing at the moment, but I would expect it to stabilize after a while, making breaking changes like this less likely.

**Comment by dacort** [comment]
@rodrigodc07 Any chance in updating this PR to accommodate the breakage mentioned by @nicolasterral?

**Comment by rodrigodc07** [comment]
College take too much time this month, and finally did it.

**Comment by srwiser** [comment]
**Unable** to load Athena driver into Metabase
@rodrigodc07 I am running Metabase 0.32.9 and have put your Athena driver jar in /plugins directory where other metabase plugins reside.
On starting up, metabase shows 
```Added URL file:/home/ubuntu/metabase/plugins/AthenaJDBC42_2.0.7.jar to classpath```
But it does not lazy load it as other plugins.

**Comment by jamroks** [comment]
Hey @dacort  is the updated code from @rodrigodc07 is enough to get merge, or is there something still holding it back ?

**Comment by dacort** [comment]
@jamroks Heyo, will take a look this weekend. Apologies for the delay.

**Comment by dacort** [comment]
@rodrigodc07 Thank you so much for putting this together - verified it works with v0.32.9 and v0.33.0-preview2. Merging now. 👍

**Pull request merged**